A refactor of the script.

dak180 opened the pull request:
The goal is to make it easier to read, more maintainable, and to solve several outstanding issues.

To do (help welcome):
- [x] new screenshots. 

Fixes #27.
Fixes #25.
Fixes #21.
Fixes #7.
Fixes #3.
Fixes #33.

fohlsso2 committed:
finally merged my local hack with the real thing

fohlsso2 committed:
finally merged my local hack with the real thing

fohlsso2 committed:
cleanup

fohlsso2 committed:
Merge branch 'master' of https://github.com/fohlsso2/FreeNAS-Report

fohlsso2 committed:
fixed scrub in progress too

fohlsso2 committed:
Update README.md

added more picture links

fohlsso2 committed:
Update README.md

added more picture links

fohlsso2 committed:
Update report.sh

fixed the parser for normal scrub

fohlsso2 committed:
Update README.md

fohlsso2 committed:
Update README.md

Commit:
Update report.sh

#   - Added Frag%, Size, Allocated, Free for ZPool status report summary.
#   - Added Disk Size, RPM, Model to the Smart Report
#   - Added if statment so that if "Model Family" is not present script will use "Device Model" 
#       for brand in the SMART Satus report details.
#   - Added Glabel Status Report (RAW)
#   - Removed Power-On time labels and added ":" as a separator.
#   - Added Power-On format to the Power-On time Header.
#   - Changed Backup deafult to false.

Commit:
Update README.md

rotx committed:
cherry-pick 66d6277

HTML boundary fix & proper message ids
Better support for NVMe and SSD
Support for new smartmon-tools

aserrallerios committed:
Rebase v1.6 to main repo

dak180 committed:
Use smart id instead of name.

Ids are standardized (mostly) but names are not.

Fixes #3.

dak180 committed:
Use a config file.

dak180 committed:
Add a check for commands.

dak180 committed:
Merge remote-tracking branch 'aserrallerios/master' into topic/temp

* aserrallerios/master:
  Rebase v1.6 to main repo
  Fixes for TrueNAS 12.0-RELEASE Adapted script because output of zpool status changed
  added line breaks to disk info table body

dak180 committed:
Shellcheck pass and general formatting.

dak180 committed:
Make sure the ssd table actually gets displayed.

dak180 committed:
Use either common wear level indicator.

dak180 committed:
Try to make things work for more ssds.

Only use smart 321 when both 231 and 233 are present.

dak180 committed:
Print NVMe smart data.

rotx committed:
Fix extended self test time parsing.

dak180 committed:
Misc updates.

dak180 committed:
Refactor to get rid of awk.

This also uses a function and is much easier to read.

dak180 committed:
A better command list.

Ordered by apearence.

dak180 committed:
Fix ups.

dak180 committed:
Do not skew the current time.

dak180 committed:
Reorg the nvme parsing.

dak180 committed:
Refactor to get rid of awk part 2.

Address the SSD table.

dak180 committed:
Refactor to get rid of awk part 3.

Address the HDD table.

dak180 committed:
More quotes.

dak180 committed:
Less sub-shells.

dak180 committed:
Put the config backup in a function

dak180 committed:
Make the zpool status into a function.

dak180 committed:
Make the UPS report into a function.

dak180 committed:
Consolidate the functions.

dak180 committed:
Add to the list of commands.

dak180 committed:
Merge branch 'master' into topic/refactor

* master:
  Correction for scans that last over 24 hours.

dak180 committed:
Use arrays in loops.

dak180 committed:
No need for nested <pre> tags.

dak180 committed:
Set arrays allowing for spaces in items.

Fixes edgarsuit/FreeNAS-Report#27

dak180 committed:
Update the ReadMe file.

dak180 committed:
Make sure the selftest info is printed properly.

dak180 committed:
Support for conveyance test output requires support from smartctl.

dak180 committed:
New preview image.

dak180 committed:
Support for conveyance test output.

dak180 committed:
Reduce calls to smartctl.

dak180 committed:
Add some notes about workarounds.

dak180 committed:
Less calls to smartctl is faster.

dak180 committed:
Standardize colorizing the rows.

Also fix indentation.

dak180 committed:
Also backup the secret seed.

dak180 committed:
Get the whole array.

dak180 committed:
correct quoting.

dak180 committed:
Fix math on NVMe write stats.

dak180 committed:
Fix units.

dak180 committed:
Support more drives.

c0dyhi11 commented:
Is it just me, or is the config file that seems to be required, not documented anywhere on what should be in it? Or how it should be formated?

dak180 commented:
@c0dyhi11 the config file will be created in a default (annotated) form on first run at the specified location if it does not already exist.

syserr0r committed:
Add Wear_Leveling_Count for some Samsung SSDs

Samsung 870-series SSDs

zanechua committed:
use system configured from name and from email

zanechua committed:
Update README

zanechua committed:
Update README about SMART tests

dak180 committed:
Fix resliver output for TrueNAS 12.

dak180 committed:
More reliable way to get LBAs written.

dak180 committed:
Account for the case of no tests run.

dak180 committed:
Update Readme to note first time configuration changes.

dak180 committed:
Merge remote-tracking branch 'syserr0r/patch-1' into topic/refactor

Fixes 2.

* syserr0r/patch-1:
  Add Wear_Leveling_Count for some Samsung SSDs

dak180 committed:
Use sane defaults for from email.

Reformat the readme.

dak180 committed:
Merge branch 'zanechua/topic/refactor' into topic/refactor

Closes 3.
* zanechua/topic/refactor:
  Update README about SMART tests
  Update README
  use system configured from name and from email

dak180 committed:
Fail unknown wear leveling gracefully.

Refs #4.

dak180 committed:
Fail unknown temp gracefully.

Refs #4.

dak180 committed:
Better quoting.

dak180 committed:
Fail unknown temp gracefully for HDDs too.

Refs #4.

dak180 committed:
Fail gracefully if the statistics page does not exist.

dak180 committed:
Account for very new drives.

Refs #4.

dak180 committed:
Adjust default drive temps.

dak180 committed:
Separate SSD temps from HDD temps.

dak180 committed:
glabel is not used on scale.

dak180 committed:
First pass at sas compatibility.

Fixes #1.

dak180 committed:
Do not care about whitespace variations.

dak180 committed:
Get rid of an extraneous %.

dak180 committed:
Make sure the line gets passed through.

dak180 committed:
Support SAS drives for detailed output.

dak180 committed:
Clean things up a bit.

dak180 committed:
Update release notes.

dak180 committed:
Accommodate both SSD and HDD temp settings.

dak180 committed:
Keep SAS drives in their own section.

dak180 committed:
Reset vars for each loop.

mth309 committed:
Update report.sh

dak180 committed:
Remove awk.

dak180 committed:
Merge branch 'topic/sas' into topic/refactor

Fixes #1.

* topic/sas:
  Remove awk.
  Update report.sh
  Reset vars for each loop.
  Keep SAS drives in their own section.
  Accommodate both SSD and HDD temp settings.
  Update release notes.
  Clean things up a bit.
  Support SAS drives for detailed output.
  Make sure the line gets passed through.
  Get rid of an extraneous %.
  Do not care about whitespace variations.
  First pass at sas compatibility.

dak180 committed:
Apparently some HDDs to not report a rotation rate so test 'Spin up Time' instead.

Fixes #7.
Closes #8.

dak180 committed:
Prevent error on nvme drives.

dak180 committed:
Colorize on test failure (HDD & SSD).

Refs #9

mth309 committed:
Update report.sh

mth309 committed:
Update report.sh

dak180 committed:
Fix white space and normalize ordering.

Fixes #9.

dak180 committed:
Remove the last awk outside of ZpoolSummary.

mth309 committed:
Update report.sh

Fixed the out-of-order check for N/A test status prior to setting that status as a workaround for improper self test support

mth309 committed:
Update report.sh

typo

mth309 committed:
Update report.sh

mth309 committed:
Fixed the parsing for when a self test is in progress on a SAS device.

The previous code was colorizing both lastTestHours and lastTestStatus as critical while the test was running due to parsing the wrong fields.

Fixes #12.

dak180 committed:
Try to be more compatible with different mail programs.

Refs #14.

dak180 committed:
More consistant sorting.

dak180 committed:
Ensure that date functions work consistently.

dak180 committed:
Fix shellcheck warnings.

dak180 committed:
More awk removal.

dak180 committed:
Use xselftest with fallback to selftest.

dak180 committed:
Make sure the blank line is where it should be.

dak180 committed:
Fix quotes.

dak180 committed:
Add examples to test against.

mkninc committed:
Add an example of long running resilver

dak180 committed:
Change to scrubDuration for better clarity and to avoid collisions.

dak180 committed:
Add sections.

dak180 committed:
Support multi-day reslivers.

Fixes #16.

dak180 committed:
Add more examples.

dak180 committed:
The last of the awk removal.

dak180 committed:
Combine multiple echos into single heredocs.

dak180 committed:
Merge branch 'topic/mail'

Fixes #14.

* topic/mail:
  Combine multiple echos into single heredocs.
  Fix quotes.
  Make sure the blank line is where it should be.
  Try to be more compatible with different mail programs.

dak180 committed:
More heredoc conversions.

dak180 committed:
Fix some quotes.

dak180 committed:
Use the extended test data if available.

dak180 committed:
Reduce the number of calls to smartctl.

dak180 committed:
Do not math non-numbers.

dak180 committed:
Account for scrubs of indeterminate and multi-day duration.

Add examples.

dak180 committed:
First pass at linux changes; untested.

dak180 committed:
Add a file dump option for debugging.

dak180 committed:
Reorder to be more consistent.

dak180 committed:
Add support for per drive overrides.

dak180 committed:
Update release notes.

dak180 committed:
Make sure to include the devstat log in the dumps.

dak180 committed:
Another command to test.

dak180 committed:
Grep out log preamble.

dak180 committed:
List NVMe self tests.

Since NVMe test support in smartctl will require version 7.4 (not yet out) use the output from nvmecontrol in the meantime.

dak180 committed:
Remove the unused -o option from sendmail.

dak180 committed:
Account for different `date` invocations.

dak180 committed:
debain date does not like _

dak180 committed:
Merge branch 'master' into topic/scale

* master: (53 commits)
  Remove the unused -o option from sendmail.
  List NVMe self tests.
  Grep out log preamble.
  Make sure to include the devstat log in the dumps.
  Update release notes.
  Add support for per drive overrides.
  Reorder to be more consistent.
  Add a file dump option for debugging.
  Account for scrubs of indeterminate and multi-day duration.
  Do not math non-numbers.
  Reduce the number of calls to smartctl.
  Use the extended test data if available.
  Fix some quotes.
  More heredoc conversions.
  Combine multiple echos into single heredocs.
  The last of the awk removal.
  Add more examples.
  Support multi-day reslivers.
  Add sections.
  Change to scrubDuration for better clarity and to avoid collisions.
  ...

dak180 committed:
Having a date field that takes almost anything is pain.

The keyword here here is almost and a lack of docs never helps.

dak180 committed:
Better drive list for scale.

dak180 committed:
nvmecontrol is bsd only.

dak180 committed:
md5sum and sha256sum are on both scale and core.

dak180 committed:
Get a more human readable status dump.

dak180 committed:
10^ is more portable than 1e.

dak180 committed:
-q does not work on debian.

dak180 committed:
Fix white space.

dak180 committed:
Fix version number.

dak180 committed:
Use current names.

dak180 committed:
Account for NVMe drives not reporting temps.

dak180 committed:
Try to handle more situations where we do not get data.

dak180 committed:
Minor edits.

Use current names.

dak180 committed:
Support SAS drives that do not support json output better.

Also note drives that do not support test monitoring through smartctl.

Refs #20.

dak180 committed:
Do not permanently set unknown values.

dak180 committed:
Fix nvme drive numbers.

dak180 committed:
Really fix nvme drive numbers.

512 because apparently the NVMe spec uses magic numbers.

dak180 committed:
Fallback for apple SSDs that do not have a stats page

dak180 committed:
Add a custom message for a missing `bc`.

dak180 committed:
Merge branch 'topic/scale'

* topic/scale:
  Add a custom message for a missing `bc`.
  Fix white space.
  -q does not work on debian.
  10^ is more portable than 1e.
  Get a more human readable status dump.
  md5sum and sha256sum are on both scale and core.
  nvmecontrol is bsd only.
  Better drive list for scale.
  Having a date field that takes almost anything is pain.
  debain date does not like _
  Account for different `date` invocations.
  Another command to test.
  First pass at linux changes; untested.
  glabel is not used on scale.

dak180 committed:
Normalize white space before comparing.

Fixes #21.

dak180 committed:
Fallback for seagate SSDs that do not have a stats page.

Refs #21.

dak180 committed:
Use single quotes where doubles are unnecessary.

Refs #22.

dak180 committed:
Fix quote.

Refs #22.

dak180 committed:
Make sure command are checked where required.

dak180 committed:
Try to restore compatibility with TrueNAS 12.

Refs #23.

dak180 committed:
Mitigate serial numbers with chars that cannot be used in bash vars.

Refs #23.

dak180 committed:
Mitigate serial numbers with chars that cannot be used in bash vars more comprehensively.

Refs #23.

dak180 committed:
Create the backup location if it does not already exist.

Fix #23.

dak180 committed:
Add a sort for linux that will work with more than 24 drives.

dak180 committed:
Allow use of `$(dirname "${configFile}")/usr/bin/` for `bc`.

This should survive updates.

Refs #6.
